Added API key column to the Tenant table

gridhead opened the pull request:
Some default answers picked for https://github.com/CentOS/duffy/issues/145#issuecomment-992082234, UUID and plain text.

Feel free to specify differently and I will amend the commits.

This forms the basis for the work on #135.

BenCapper reviewed (approved):
LGTM

nphilipp commented on duffy/api_models/tenant.py:
```diff
@@ -11,6 +11,7 @@
 class TenantBase(BaseModel, ABC):
     name: str
     is_admin: Optional[bool]
+    api_key: str
```
As discussed, I don't think we want to return this in queries (regardless of if the API key is stored hashed or not), so let's move it to `TenantCreateModel`, and use [`pydantic.UUID4`](https://pydantic-docs.helpmanual.io/usage/types/#pydantic-types) as the type (which validates that the supplied string is a version 4 – i.e. random – UUID and converts it to an equivalent `uuid.UUID` object .

gridhead replied:
Addressed here https://github.com/CentOS/duffy/pull/146/commits/db161b13c2ef2d300d65dd708281961e677b5998

nphilipp commented on duffy/database/model/tenant.py:
```diff
@@ -10,4 +10,5 @@ class Tenant(Base, CreatableMixin, RetirableMixin):
     id = Column(Integer, primary_key=True, nullable=False)
     name = Column(UnicodeText, unique=True, nullable=False)
     is_admin = Column(Boolean, nullable=False, default=False, server_default=text("FALSE"))
+    api_key = Column(UnicodeText, nullable=False)
```
I guess this could be a simple `Text`, and we probably want to proxy this a little: setting `api_key` to a `uuid.UUID()` value would hash the value using e.g. [bcrypt](https://pypi.org/project/bcrypt/), but reading it would simply return the hashed value (which the Pydantic model filters out). Verification would be done by e.g. a `validate_api_key()` method which we'd use in the controllers. Could look somewhat like this (note: leading underscore of the "raw" class attribute):

```
    import uuid
...
    from pydantic import UUID4
...
   from sqlalchemy.ext.hybrid import hybrid_property
...
    _api_key = Column("api_key", Text, nullable=False)
...
    @hybrid_property
    def api_key(self):
        return self._api_key

    @api_key.setter
    def api_key(self, key: uuid.UUID):
        salt = bcrypt.gensalt()
        hashed_key = bcrypt.hashpw(key.bytes, salt)
        # bcrypt uses binary strings
        self._api_key = hashed_key.decode("ascii")

    def validate_api_key(self, key: uuid.UUID):
        return bcrypt.checkpw(key.bytes, self._api_key.encode("ascii"))
```

gridhead replied:
Addressed here https://github.com/CentOS/duffy/pull/146/commits/db161b13c2ef2d300d65dd708281961e677b5998

nphilipp commented on tests/app/controllers/test_tenant.py:
```diff
@@ -9,6 +9,7 @@ class TestTenant(BaseTestController):
     path = "/api/v1/tenants"
     attrs = {
         "name": "Some Honky Tenant!",
+        "api_key": "2d4b9f2b-ad45-484f-a0a6-75393bacba4c",
```
With the above change, this line (here and below) would have to become something like that:

```
import uuid
...
        "api_key": uuid.uuid4(),
```

gridhead replied:
Addressed here https://github.com/CentOS/duffy/pull/146/commits/d99d8ecb7c3f0694da61e9c2a827e35a0f0a5c32

gridhead committed:
Add API key to tenant model and endpoints

The Tenant attribute `api_key` transparently hashes the key with bcrypt
when set. Following, we have to skip validating its contents when
testing the model or the associated endpoints.

Signed-off-by: Akashdeep Dhar <akashdeep.dhar@gmail.com>

gridhead commented:
To https://github.com/CentOS/duffy/pull/146#issue-1078080433, in the subsequent reworks on the pull request - we settled on the answers UUID and masking using bcrypt.

nphilipp commented on tests/app/controllers/__init__.py:
```diff
@@ -102,15 +102,16 @@ def _verify_item(cls, item, attrs=None):
-                    # skip verifying objects this one depends on
-                    continue
-            assert item[key] == value
+            if key != "api_key":
+                try:
+                    subcls, _ = cls.attrs[key]
+                except (TypeError, ValueError):
+                    pass
+                else:
+                    if isinstance(subcls, type) and issubclass(subcls, BaseTestController):
+                        # skip verifying objects this one depends on
+                        continue
+                assert item[key] == value
```
Uhh, we shouldn't special case things like this in `BaseTestController`. I have a similar requirement to have fields submitted in POST that shouldn't be verified in the result, let me carve that out of the heap of changes in my branch and put it in here.

nphilipp replied:
As discussed, there's `BaseTestController.no_response_attrs` now which can be set in child classes (which I did with the latest push).

nphilipp commented on tests/database/test_model.py:
```diff
@@ -38,13 +46,19 @@ async def test_query_obj_async(self, db_async_obj):
+            if key == "api_key":
+                assert validate_api_key(value, obj.api_key)
+            else:
+                objvalue = getattr(obj, key)
+                if isinstance(objvalue, (int, str)):
+                    assert objvalue == value
         for key, value in self._db_obj_get_dependencies().items():
-            objvalue = getattr(obj, key)
-            if isinstance(objvalue, (int, str)):
-                assert objvalue == value
+            if key == "api_key":
+                assert validate_api_key(value, obj.api_key)
+            else:
+                objvalue = getattr(obj, key)
+                if isinstance(objvalue, (int, str)):
+                    assert objvalue == value
```
Similar things as with `BaseTestController`, let's not special case in the base test class.

nphilipp replied:
Same thing as above, with `ModelTestBase.no_validate_attrs`.

nphilipp commented on duffy/database/model/tenant.py:
```diff
@@ -10,4 +14,20 @@ class Tenant(Base, CreatableMixin, RetirableMixin):
+        self._api_key = hashed_key.decode("ascii")
+
+
+def validate_api_key(key: str, hash: str):
+    return bcrypt.checkpw(str(key).encode("ascii"), hash.encode("ascii"))
```
I'd rather have this as a method of `Tenant` objects (with its own test):

```
    def api_key(...):
        ...

    def validate_api_key(self, key: uuid.UUID) -> bool:
        return bcrypt.checkpw(str(key).encode("ascii"), self.api_key.encode("ascii"))
```

Would be used somehow like this:

```
    result = await DBSession.execute(select(Tenant).filter_by(...))
    tenant = result.scalar_one_or_none()
    if not (tenant and tenant.validate_api_key(key)):
        raise HTTPException(...)
```

gridhead replied:
I have restored the `validate_api_key` (previously) function as a method under the `Tenant` class but I did not add the latter parts as they are more involved in the authentication which is beyond the scope of this pull request. I'd address those when I'll get to work on a PR to fix #135.

gridhead replied:
Ok, scratch whatever I said at https://github.com/CentOS/duffy/pull/146#discussion_r770183711. I would need to implement the latter parts as well so as to write the tests for the `Tenant` object - even though the PR is not about it :/

nphilipp commented on tests/database/test_model.py:
```diff
@@ -68,7 +83,11 @@ class TestSession(ModelTestBase):
-        tenant = model.Tenant(name="My Other Tenant", ssh_key="my other public SSH key")
+        tenant = model.Tenant(
+            name="My Other Tenant",
+            api_key=str(uuid4()),
```
`api_key` should accept `UUID` objects, there should be no need to cast it to `str` here and below.

nphilipp replied:
Latest push fixes this.

nphilipp committed:
Make API attribute testing more flexible

Previously, all attributes submitted to an endpoint in POST were
expected to be present in the response with the exception of items
proxied to another controller test class for dependencies.

This lets subclasses set `no_response_attrs` to a list of attribute
names which aren't expected in the response and thus won't be verified.

Signed-off-by: Nils Philippsen <nils@redhat.com>

nphilipp committed:
Make DB model attribute validation more flexible

Previously, all attributes used when creating an object were compared to
the attributes on the object. This lets us add names of attributes where
this shouldn't be done to no_validate_attrs of the model test class.

Signed-off-by: Nils Philippsen <nils@redhat.com>

nphilipp reviewed (changes requested):
As discussed, I slipped the changes to `BaseTestController` and `ModelTestBase` underneath your commits which allow marking attributes not to be validated and adapted your commits to fit. Other than the other inline comments, please squash your three commits as they form one logical unit: new code, its dependencies and tests. Thanks!
